### ISSUE TITLE
shellcheck can't parse actual bash function name

### ISSUE THREAD

**Issue by ronburk** [koalaman/shellcheck#909]
#### For bugs
- Rule Id SC1089
- My shellcheck version 0.4.6
- [ ] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh
# can we really do funky crap like this?
function [}{*(){
   echo "Let me take you to, funkytown!"
}
[}{*    # why yes, we can!
# though editor auto-indent modes may punish us

```

#### Here's what shellcheck currently says:
emits an SC1089 on the trailing '}' of function definition

#### Here's what I wanted or expected to see:
Oh, I don't know. Maybe something like "bash allows nearly every character (printable or not) as a function name, but I'm just going to quit now if you're going to write that sort of code."

Inspired by seeing actual function names containing '?' in the wild. Documented more here:
http://stackoverflow.com/questions/28114999/what-are-the-rules-for-valid-identifiers-e-g-functions-vars-etc-in-bash/44041384#44041384

**Comment by firegurafiku** [comment]
@koalaman

Let me rise your attention to this issue. Of course, nobody is going to name their functions like `[}{*` in real life, but is it possible to add some more characters into the set of ones allowed in bash function names? Bash is fine with `/`, `@`, `^`, `%`, `*`, and `=`, but they're making ShellCheck stop further parsing. The first four of them are relatively safe to allow.

**Comment by koalaman** [comment]
Interestingly this only applies when using the `function` keyword. It does not work for normal function definitions.

**Comment by koalaman** [comment]
Actually that was just when the () happened to turn into part of an extglob, e.g. `foo+() { bar; }`

I've added most of these characters, but `[]!` is only recognized when using `function` to avoid misdiagnosis of `[` issues, and `{}` is not recognized at all because `function foo{bar` is far more likely to be missing spaces than a valid name.

**Comment by aadrian** [comment]
> ... Of course, nobody is going to name their functions like [}{*

Maybe https://github.com/niieani/bash-oo-framework#using-try--catch  ?

Rule SC1089 will break when using the above try/catch pattern :( .
